rest/x509_utils_test.go

```diff
@@ -0,0 +1,310 @@
+	respBody, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("couldn't reloadCertificate: expected %d status code but got %d: %s", http.StatusOK, resp.StatusCode, respBody)
+	}
+	base.Debugf(base.KeyAll, "triggered reload of certificates on Couchbase Server")
+
+	if err := enableX509ClientCertsInCouchbaseServer(restAPIURL); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func enableX509ClientCertsInCouchbaseServer(restAPIURL url.URL) error {
+	clientAuthSettings := bytes.NewBufferString(`{
+  "state": "enable",
+  "prefixes": [{
+    "path": "subject.cn",
+    "prefix": "",
+    "delimiter": ""
+}]
+}`)
+
+	// Configure CBS to enable optional X.509 client certs
+	resp, err := http.Post(restAPIURL.String()+"/settings/clientCertAuth", "application/json", clientAuthSettings)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusAccepted {
+		return fmt.Errorf("couldn't configure clientCertAuth: expected %d or %d status codes but got %d: %s", http.StatusOK, http.StatusAccepted, resp.StatusCode, respBody)
+	}
+	base.Debugf(base.KeyAll, "enabled X.509 client certs in Couchbase Server")
+
+	return nil
+}
+
+// sshCopyFileAsExecutable takes in a full source filepath, an SSH remote host (e.g. root@localhost), and a destination directory.
+// The destination directory will be created in full if it does not exist, the file will be copied, and then the read and execute permissions set.
+func sshCopyFileAsExecutable(sourceFilepath, sshRemoteHost, destinationDirectory string) error {
+	const (
+		// SSH option flags
+		forceKeyOnly        = "-o PasswordAuthentication=no"
+		skipHostFingerprint = "-o StrictHostKeyChecking=no"
+	)
+
+	// make destination directory if it doesn't exist
+	cmd := exec.Command("ssh", forceKeyOnly, skipHostFingerprint, sshRemoteHost, "mkdir", "-p", destinationDirectory)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return errors.Wrap(err, string(output))
+	}
+
+	// copy the file (requires SSH Keys to be set up)
+	cmd = exec.Command("scp", forceKeyOnly, skipHostFingerprint, sourceFilepath, sshRemoteHost+":"+destinationDirectory)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return errors.Wrap(err, string(output))
+	}
+
+	// make the file we just copied readable and executable (required for Couchbase server to use the certs)
+	cmd = exec.Command("ssh", forceKeyOnly, skipHostFingerprint, sshRemoteHost, "chmod -R a+rx", filepath.Join(destinationDirectory, filepath.Base(sourceFilepath)))
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return errors.Wrap(err, string(output))
+	}
+
+	return nil
+}
+
+// domainForIPAddr returns a domain name that publicly resolves to the given IP address.
+func domainForIPAddr(ip net.IP) string {
```

Comment:
This is a nice trick!  Does it add any requirements for the test environment that we need to worry about (e.g. public internet access)?

Reply:
Oops, this was used when I was automatically renaming the cluster node, but not required any more... Still using xip.io for testing this, but manually instead.

It does require public internet access to resolve them. For completely isolated test environments I'd imagine we can use /etc/hosts or IT's internal DNS to provide local-only test domains